src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php

```diff
@@ -230,114 +190,106 @@ public function putAction($id, Request $request)
+     * for children of a category.
+     * If expandIds is set, the paths to the categories which are assigned to the ids are expanded. Categories are
+     * only expanded if they are descendants of the root level and the do not affect the limit or pagination.
      *
-     * @throws MissingParameterException iff the locale is missing in the request
+     * @param Request $request
+     * @param $locale
+     * @param null $parentId
+     * @param bool $disableLimit
+     * @param array $expandIds
      *
      * @return CategoryListRepresentation
      */
-    protected function getCategoryListRepresentation(Request $request, $parentKey = null)
+    protected function getListRepresentation(Request $request, $locale, $parentId = null, $disableLimit = false, $expandIds = [])
     {
-        /** @var RestHelperInterface $restHelper */
-        $restHelper = $this->getRestHelper();
-
-        /** @var DoctrineListBuilderFactory $factory */
-        $factory = $this->get('sulu_core.doctrine_list_builder_factory');
+        $listBuilder = $this->initializeListBuilder($locale);
 
-        $listBuilder = $factory->create(self::$entityName);
-        $locale = $this->getRequestParameter($request, 'locale', true);
-        $fieldDescriptors = $this->getManager()->getFieldDescriptors($locale);
-        $listBuilder->sort($fieldDescriptors['depth']);
+        // if search is set, return all matching categories. else return only level below parent in category tree
+        if (!$request->get('search')) {
+            $listBuilder->where($this->getCategoryManager()->getFieldDescriptor($locale, 'parent'), $parentId);
+        }
 
-        $restHelper->initializeListBuilder($listBuilder, $fieldDescriptors);
+        // need to disable default limit because some frontend components are not paginated
+        if ($disableLimit || !$request->get('limit')) {
+            $listBuilder->limit(null);
+        }
 
-        $listBuilder->addSelectField($fieldDescriptors['depth']);
-        $listBuilder->addSelectField($fieldDescriptors['parent']);
-        $listBuilder->addSelectField($fieldDescriptors['locale']);
-        $listBuilder->addSelectField($fieldDescriptors['defaultLocale']);
-        $listBuilder->addSelectField($fieldDescriptors['lft']);
-        $listBuilder->addSelectField($fieldDescriptors['rgt']);
+        $results = $listBuilder->execute();
 
-        if ($parentKey !== null) {
-            $this->addParentSelector($parentKey, $listBuilder);
+        if ($expandIds) {
+            // expand the paths to the categories which are assigned to the given expandIds.
+            // we need to do this in a second query, because we need to expand children of the current page only.
+            $resultIds = array_map(
+                function ($category) {
+                    return $category['id'];
+                },
+                $results
```

Comment:
the result of the first listbuilder instance is **not** used for the ids only. this result contains the current page/searchresult of the request. 
If an expandIds flag is set, we need to expand the paths to this ids additionally. This is done with a second listbuilder instance.  The second instance only loads the path to the ids, nothing is loaded twice. The two results are merged.

Using two different listbuilder instances is not the perfect way, but replacing the first listbuilder instance with a repository-method means reimplementing pagination.
Pagination in this case is quite complicated, as it is dependent of the sorting, which is again dependent on the translation.